### PR TITLE
Improve ClickHouse data type mapping documentation

### PR DESCRIPTION
**Pull request by tlblessing**
## Description

Add table for reverse mapping direction. Clean up table formatting.

image

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Connector documentation

> How would you describe this change to a non-technical end user or system administrator?

Added a table of the data types from Trino back to ClickHouse. Reformatted table for better wrapping and consistency.

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

**Review by ebyhr — COMMENTED** [review]
There's an existing PR #13259. Could you coordinate with @tangjiangling ?

**Comment by tlblessing** [comment]
> There's an existing PR #13259. Could you coordinate with @tangjiangling ?

Sure, I could apply your review comments to my PR. Is that okay with you @tangjiangling ?

cc: @jhlodin

**Comment by tangjiangling** [comment]
> Is that okay with you @tangjiangling ?

Sure.

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 1]
```diff
@@ -208,15 +208,15 @@ ClickHouse        Trino             Notes
...
+
 
 Trino type to ClickHouse type mapping
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 The connector maps Trino types to the corresponding ClickHouse types following
 this table:
 
-
+.. include:: jdbc-type-mapping.fragment
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:336]: Please squash this commit to "Add headings for bi-directional tables".

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 2]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+The connector maps Trino types to the corresponding ClickHouse types according
+to the following table:
+
+.. list-table:: Trino type to ClickHouse type mapping
+  :widths: 30, 20, 50
+  :header-rows: 1
+
+  * - Trino type
+    - ClickHouse type
+    - Notes
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:270]: Add boolean mapping:
```rst
  * - ``BOOLEAN``
    - ``UInt8``
    -
```

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 3]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    - ``BIGINT`` is an alias of ``Int64``
+  * - ``SMALLINT``
+    - ``UInt8``
+    -
+  * - ``INTEGER``
+    - ``UInt16``
+    -
+  * - ``BIGINT``
+    - ``UInt32``
+    -
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:291]: Remove duplicated entries.

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 4]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+  * - ``DECIMAL(20,0)``
+    - ``UInt64``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:293]: Remove as duplicate of subsequent entries.

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 5]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    -
+  * - ``REAL``
+    - ``Float32``
+    - ``FLOAT`` is an alias of ``Float32``
+  * - ``DOUBLE``
+    - ``Float64``
+    - ``DOUBLE`` is an alias of ``Float64``
+  * - ``DECIMAL``
+    - ``Decimal``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:302]:
```suggestion
  * - ``DECIMAL(p,s)``
    - ``Decimal(p,s)``
```

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 6]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    -
+  * - ``VARBINARY``
+    - ``FixedString``
+    - Enabling ``clickhouse.map-string-as-varchar`` config property changes the
+      mapping to ``VARCHAR``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:307]: Remove as duplicate of subsequent entries.

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 7]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+  * - ``VARBINARY``
+    - ``String``
+    - Enabling ``clickhouse.map-string-as-varchar`` config property changes the
+      mapping to ``VARCHAR``
+  * - ``DATE``
+    - ``Date``
+    -
+  * - ``TIMESTAMP``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:315]:
```suggestion
  * - ``TIMESTAMP(0)``
```

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 8]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    - ``DateTime``
+    -
+  * - ``IPADDRESS``
+    - ``IPv4``
+    -
+  * - ``IPADDRESS``
+    - ``IPv6``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:322]: Remove as the connector doesn't support writing `IPADDRESS` type.

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 9]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    -
+  * - ``VARCHAR``
+    - ``Enum8``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:325]: Move to before `VARBINARY`. 
```suggestion
  * - ``VARCHAR``
    - ``String``
```

File: docs/src/main/sphinx/connector/clickhouse.rst [hunk 10]
```diff
@@ -172,32 +172,166 @@ in create table statement. ``ReplicatedMergeTree`` engine is not yet supported.
...
+    -
+  * - ``VARCHAR``
+    - ``Enum16``
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/clickhouse.rst:328]: Move to before `VARBINARY`.
```suggestion
  * - ``CHAR``
    - ``String``
```

**Review by ebyhr — COMMENTED** [review]
I would recommend improving the commit titles to clarify this is document change. It helps us when reading commit history. 

* Add header for ClickHouse type mapping documentation
* Document write mapping and reformat table in ClickHouse
* Fix typo in ClickHouse documentation

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 2 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:270]: I added it with the other UIntx options line 283

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 3 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:291]: These are not duplicate entries.

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 8 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:322]: Done

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 4 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:296]: ?

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 6 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:310]: ?

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 3 above)
Comment (ebyhr) [reply docs/src/main/sphinx/connector/clickhouse.rst:294]: * SMALLINT already exists in L274
* INTEGER already exists in L277
* BIGINT already exists in L280

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 4 above)
Comment (ebyhr) [reply docs/src/main/sphinx/connector/clickhouse.rst:296]: DECIMAL already exists in L304

**Comment by colebow** [comment]
You're going to need to squash these later commits down into the proper original commits. Trino has a strict policy on commits being atomic, sensible units of change, and it's similarly strict with commit messages, and we try to follow this guide: https://cbea.ms/git-commit/

When you get feedback or suggestions, you should amend your edits into the commit where you made the change.

My workflow for this is usually:

Make my changes, stage my changes
`git log -3` (or whatever number you need)
Copy the hash of the commit I'm editing
`git commit --fixup <hash>`
`git rebase -i HEAD~3 --autosquash`

And that should get all changes into the original commit as appropriate. There's several different ways to accomplish the same thing in Git, but that strategy works best for me.

If you've already made the edit commit like you have here, you can do a rebase, re-order the unnecessary commits to follow the ones they're editing, and change them from `pick` to `fixup` to merge them into the proper commits.

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 3 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/clickhouse.rst:294]: Agreed with @ebyhr since that logically reflects what is happening in the type mapping effort

**Comment by mosabua** [comment]
@tlblessing and @colebow please work on the git-fu together as necessary.

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 4 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst:296]: Will do but they look different to me 
image

Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst]: Done in commit [0eaf1e9](https://github.com/trinodb/trino/pull/13578/commits/0eaf1e9a0ceb3f713e4915108ec65370c8e257c7)

**Comment by tlblessing** [comment]
> @tlblessing and @colebow please work on the git-fu together as necessary.

Will do. Cole is OOO until mid next week.

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 3 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst]: Done in commit [c40e96c](https://github.com/trinodb/trino/pull/13578/commits/c40e96cec6200806d4a416c2f5deb86b67c2228e)

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 6 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst]: Done in commit [d563ed5](https://github.com/trinodb/trino/pull/13578/commits/d563ed5b5935c6b67bc57e98b11d86cc474582db)

**Comment by tlblessing** [comment]
I assume that these types should also be all uppercase @ebyhr

File: docs/src/main/sphinx/connector/clickhouse.rst (hunk 10 above)
Comment (tlblessing) [reply docs/src/main/sphinx/connector/clickhouse.rst]: Done in commit [b66dcf0](https://github.com/trinodb/trino/pull/13578/commits/b66dcf00beeca1c12983d002362d0d7ce6526253)

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Merged, thanks!

**Pull request merged**